awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -440,53 +459,113 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
-                    if 'error' in response['json']:
-                        self.fail_json(msg="Unable to delete {0} {1}: {2}".format(item_type, item_name, response['json']['error']))
+                self.fail_json(msg="Unable to delete {0} {1}: {2}".format(item_type, item_name, self.extract_errors_from_response(response)), **{'response': response})
+        else:
+            self.exit_json(**self.json_output)
+
+    def modify_associations(self, association_endpoint, new_association_list):
+        # if we got None instead of [] we are not modifying the association_list
+        if new_association_list == None:
+            return
+
+        # First get the existing associations
+        response = self.get_all_endpoint(association_endpoint)
+        existing_associated_ids = [association['id'] for association in response['json']['results']]
+
+        # Disassociate anything that is in existing_associated_ids but not in new_association_list
+        ids_to_remove = list(set(existing_associated_ids) - set(new_association_list))
+        for an_id in ids_to_remove:
+            response = self.post_endpoint(association_endpoint, **{'data': {'id': int(an_id), 'disassociate': True}})
+            if response['status_code'] == 204:
+                self.json_output['removed_an_association'] = True
+                self.json_output['changed'] = True
+            else:
+                self.fail_json(msg="Failed to disassociate item {0}".format(self.extract_errors_from_response(response)))
+
+        # Associate anything that is in new_association_list but not in `association`
+        for an_id in list(set(new_association_list) - set(existing_associated_ids)):
+            response = self.post_endpoint(association_endpoint, **{'data': {'id': int(an_id)}})
+            if response['status_code'] == 204:
+                self.json_output['added_an_association'] = True
+                self.json_output['changed'] = True
+            else:
+                self.fail_json(msg="Failed to associate item {0}".format(self.extract_errors_from_response(response)))
+
+    def process_additional_posts(self, item, item_type, additional_posts):
+        # Process any additional posts with this item
+        for post in additional_posts:
+            endpoint = item.get('related', {}).get(post['endpoint_reference'], None)
+            if endpoint:
+                # First get the existing data from the endpoint
+                existing_data = self.get_endpoint(endpoint)
+
+                # If the endpont data is identical, just continue on to the next item
+                if existing_data['json'] == post['data']:
+                    continue
+
+                with open('/tmp/john', 'w') as f:
+                    import json
+                    f.write(json.dumps(post['data']))
+
+                # If this item no longer has data, lets remove it
+                if post['data'] in ('', [], {}):
```

Comment:
This could cause weird changed=True cases, since `existing_data['json'] == post['data']` will be unaware of cases which are different in syntax, but both indicate various version of nothing.

Reply:
Do you have examples of that? This was a first crack at defining some kind of pattern, if we can't make it work we can try something else.

Reply:
So I guess I'm not following this at all. This `if` should be written as `if not post['data']` unless I'm missing something. If `if 'data' in post and (not post['data'])`. Unless the intent is that `None` should not result in a DELETE. This is why the contract gets really hard to manage. There's a lot of code here for survey spec handling, I don't think it generalizes well, it doesn't align with WFJT surveys, and I would go back to the drawing board for everything with the spec.

Reply:
Correct, None would represent that the survey_spec (or any other field) was just not provided to the module. So we should not delete the survey on a None condition. The module itself is currently catching survey_spec (or any other field) being None and choosing to not add the data to the additional post parameter. However we could also go the route of just passing None back into this method (especially if we are doing auto-generation).  So instead of blocks like this:
```
   additional_posts = []
    if survey_spec is not None:
        additional_posts.append({
            'endpoint_reference': 'survey_spec',
            'data': survey_spec,
        })
```
Followed with:
```
        module.create_or_update_if_needed(
            existing_item, new_fields,
            endpoint='job_templates', item_type='job_template',
            associations={
                'credentials': credentials_ids,
                'labels': labels_ids,
            },
            additional_posts=additional_posts,
        )
```
We could just do something like:
```
        module.create_or_update_if_needed(
            existing_item, new_fields,
            endpoint='job_templates', item_type='job_template',
            associations={
                'credentials': credentials_ids,
                'labels': labels_ids,
            },
            additional_posts=[
              {'endpoint_reference': 'survey_spec','data': survey_spec,}
            ],
        )
```

It looks like survey_spec is the same for WJT as it is for JT (unless of course I'm missing something) its even posted through the same survey_spec end point. I did a quick test converting WFJT module to use additional_posts and it seemed to work (granted it wasn't a very large survey).